File: julia/algorithms/hedge/hedge/hedge.jl
type Hedge <: BanditAlgorithm
  eta::Float64
  counts::Vector{Int64}
  values::Vector{Float64}
end

function Hedge(n_arms::Int64, eta::Float64)
  Hedge(eta, zeros(Int64, n_arms), zeros(n_arms))
end

function initialize(algo::Hedge, n_arms::Int64)
  algo.counts = zeros(Int64, n_arms)
  algo.values = zeros(n_arms)
end

function initialize(algo::Hedge, arms::Vector)
  initialize(algo, length(arms))
end

function select_arm(algo::Hedge)
  z = sum(exp(algo.values * algo.eta))
  probs = exp(algo.values * algo.eta) ./ z
  return rand(Categorical(probs))
end

function update(algo::Hedge, chosen_arm::Int64, reward::Real)
  algo.counts[chosen_arm] += 1

  value = algo.values[chosen_arm]
  algo.values[chosen_arm] = value + reward
end

File: python/algorithms/hedge/hedge.py
import math
import random

def categorical_draw(probs):
  z = random.random()
  cum_prob = 0.0
  for i in range(len(probs)):
    prob = probs[i]
    cum_prob += prob
    if cum_prob > z:
      return i
  
  return len(probs) - 1

class Hedge:
  def __init__(self, temperature, counts, values):
    self.temperature = temperature
    self.counts = counts
    self.values = values
    return
  
  def initialize(self, n_arms):
    self.counts = [0 for col in range(n_arms)]
    self.values = [0.0 for col in range(n_arms)]
    return
  
  def select_arm(self):
    try:
      z = sum([math.exp(v / self.temperature) for v in self.values])
    except OverflowError:
      z = math.inf
    probs = []
    for v in self.values:
      try:
        probs.append(math.exp(v / self.temperature) / z)
      except OverflowError:
        probs.append(math.inf)
    return categorical_draw(probs)
  
  def update(self, chosen_arm, reward):
    self.counts[chosen_arm] = self.counts[chosen_arm] + 1
    
    value = self.values[chosen_arm]
    self.values[chosen_arm] = value + reward
    return
